src/Parallafka/CommitState.cs

```diff
@@ -0,0 +1,185 @@
+                    return false;
+                }
+            }
+            finally
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitReadLock();
+            }
+
+            lock (messagesNotYetCommitted)
+            {
+                while (messagesNotYetCommitted.TryPeek(out IKafkaMessage<TKey, TValue> msg) && 
+                       msg.WasHandled && 
+                       msg.Offset.Offset <= message.Offset.Offset)
+                {
+                    if (messageToCommit == null || msg.Offset.Offset > messageToCommit.Offset.Offset)
+                    {
+                        messageToCommit = msg;
+                    }
+
+                    messagesNotYetCommitted.Dequeue();
+
+                    this._canQueueMessage.Release();
+
+                    Parallafka<TKey, TValue>.WriteLine($"CS:DequeueMessage: {msg.Key} {msg.Offset}");
+                }
+
+                Parallafka<TKey, TValue>.WriteLine($"CS:GetMsgToCommit: {message.Key} {message.Offset} {(messageToCommit == null ? "[notfound]" : messageToCommit.Offset )}");
+
+                return messageToCommit != null;
+            }
+        }
+
+        /// <summary>
+        /// Adds the message to the commit queue to eventually be committed
+        /// </summary>
+        /// <param name="message"></param>
+        public async Task EnqueueMessageAsync(IKafkaMessage<TKey, TValue> message)
+        {
+            Queue<IKafkaMessage<TKey, TValue>> messagesInPartition;
+
+            this._messagesNotYetCommittedByPartitionReaderWriterLock.EnterReadLock();
+            try
+            {
+                this._messagesNotYetCommittedByPartition.TryGetValue(message.Offset.Partition, out messagesInPartition);
+            }
+            finally
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitReadLock();
+            }
+
+            if (messagesInPartition == null)
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.EnterWriteLock();
+                try
+                {
+                    if (!this._messagesNotYetCommittedByPartition.TryGetValue(message.Offset.Partition, out messagesInPartition))
+                    {
+                        messagesInPartition = new();
+                        this._messagesNotYetCommittedByPartition[message.Offset.Partition] = messagesInPartition;
+                    }
+                }
+                finally
+                {
+                    this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitWriteLock();
+                }
+            }
+
+            // ReSharper disable once InconsistentlySynchronizedField
+            await this._canQueueMessage.WaitAsync(-1, this._stopToken);
+
+            lock (messagesInPartition)
+            {
+                messagesInPartition.Enqueue(message);
```

Comment:
I considered ConcurrentQueue, but we need to lock outside of actual queue manipulation - we're peeking the queue and then doing something.

ConcurrentQueue is not faster than lock(Queue) - they're the same thing.

The queue is accessed by multiple threads/tasks:
. the kafka poller
. the periodic timer to commit when a message is stuck
. the committer after a message is completed.

Reply:
ConcurrentQueue doesn't lock on Enqueue unless it has to add segments. Happy path: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Concurrent/ConcurrentQueueSegment.cs#L269

_Enqueue_ should only be called by one thread at the moment since we have one poller.

Most importantly, if we use ConcurrentQueue, Enqueue() calls don't contend with the locks to guard GetMessageToCommit queue logic.

Reply:
Eliminating the double readers allows us to switch to ConcurrentQueue, see https://github.com/Snagajob/parallafka-for-dotnet/pull/1